### PR TITLE
Add test automation for Alpine

### PR DESCRIPTION
**Pull request by halter73**
I manually triggered the aspnetcore-helix-matrix, so I think it should show up in the PR checks.

Addresses #15277

### CODE REVIEW

File: src/ProjectTemplates/test/GrpcTemplateTest.cs [hunk 1]
```diff
@@ -33,7 +33,7 @@ public ITestOutputHelper Output
         }
 
         [ConditionalFact]
-        [SkipOnHelix("Not supported queues", Queues = "Windows.7.Amd64;Windows.7.Amd64.Open;Windows.81.Amd64.Open;All.OSX")]
+        [SkipOnHelix("Not supported queues", Queues = "Windows.7.Amd64;Windows.7.Amd64.Open;Windows.81.Amd64.Open;All.OSX;(Alpine.312.Amd64.Open)Ubuntu.1604.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.12-helix-20200908125345-56c6673")]
```
Comment (HaoK) [inline src/ProjectTemplates/test/GrpcTemplateTest.cs:36]: Should we future proof this now and introduce/generalize the alpine to be similar to All.OSX skip?

Comment (HaoK) [reply src/ProjectTemplates/test/GrpcTemplateTest.cs:36]: i.e. more cases or more smarts to https://github.com/dotnet/aspnetcore/blob/master/src/Testing/src/xunit/SkipOnHelixAttribute.cs#L54 and potentially https://github.com/dotnet/aspnetcore/blob/master/eng/targets/Helix.targets#L100 for All support in SkipHelixQueues

**Comment by Tratcher** [comment]
Can you summarize here for posterity why are each of these test projects is being skipped?

**Comment by halter73** [comment]
> Can you summarize here for posterity why are each of these test projects is being skipped?

For libuv tests, we just don't have a libuv build that targets Alpine and it's now obsolete. I'm not sure about the specifics for the others, but in every case we're already skipping on ARM which usually means there's some custom binary outside of the runtime in play. AFAIK, it's not really possible to target both glibc and musl in a single binary, so it makes sense that tests that cannot run on ARM also cannot run on Alpine.

File: src/SignalR/clients/java/signalr/test/signalr.client.java.Tests.javaproj [hunk 2]
```diff
@@ -6,6 +6,7 @@
     <IsTestProject>true</IsTestProject>
     <!-- Installing Java on ARM will take some work -->
     <SkipHelixArm>true</SkipHelixArm>
+    <SkipHelixQueues>(Alpine.312.Amd64.Open)Ubuntu.1604.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.12-helix-20200908125345-56c6673</SkipHelixQueues>
```
Comment (BrennanConroy) [inline src/SignalR/clients/java/signalr/test/signalr.client.java.Tests.javaproj:9]: Did you look into why this one failed? I assume the java bits we install will work on alpine

Comment (halter73) [reply src/SignalR/clients/java/signalr/test/signalr.client.java.Tests.javaproj:9]:
```
PlatformArch: linux-x64
Downloading from: https://netcorenativeassets.blob.core.windows.net/resource-packages/external/linux/java/jdk-10.0.2_linux-x64_bin.tar.gz
Installing java from jdk-10.0.2_linux-x64_bin.tar.gz https://netcorenativeassets.blob.core.windows.net/resource-packages/external/linux/java/jdk-10.0.2_linux-x64_bin.tar.gz
Unpacking to /root/helix/work/workitem/java
/root/helix/work/workitem
+ '['  '='  ]
+ export 'JAVA_HOME=/root/helix/work/workitem/java'
+ chmod +x ./gradlew
+ ./gradlew '-Dorg.gradle.daemon=false' '-PpackageVersion=6.0.0-ci' test
./gradlew: exec: line 183: /root/helix/work/workitem/java/bin/java: not found
```

https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-aspnetcore-refs-heads-halter73-15277-44d69a8221e240158c/signalr.client.java.Tests--/console.173bce24.log?sv=2019-07-07&se=2020-11-12T22%3A27%3A23Z&sr=c&sp=rl&sig=e5XON14ymqsnY%2B0Uz4Aw7tFRUCcanUqbVmZEHJfIZ7s%3D

https://dev.azure.com/dnceng/public/_build/results?buildId=864541&view=results

Why would you assume the java bits would work on alpine? We already skip this on ARM.

Comment (BrennanConroy) [reply src/SignalR/clients/java/signalr/test/signalr.client.java.Tests.javaproj]: Alpine is just a linux distro right? It's still x64 and should work the same as all the other linux distro's we run on.

From the error, it looks like either tar didn't work correctly, or the scripting didn't work https://github.com/dotnet/aspnetcore/blob/219ecd6880120215137f4d6ef0dd2aa62c780af2/eng/targets/Helix.targets#L4

**Comment by halter73** [comment]
Can someone more proficient with msbuild explain why I'm not getting the following error?

```
Expected a property at position 1 in condition "$(HelixTargetQueue.StartsWith('(Alpine.')) AND '$(SkipHelixAlpine)' == 'true'". Did you forget the closing parenthesis?
```

I don't think I forgot a closing parenthesis. Is the '(' inside the StartsWith parameter causing this somehow?

**Comment by BrennanConroy** [comment]
Likely the `'` is illegal, https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-special-characters?view=vs-2019

File: src/SignalR/clients/java/signalr/test/signalr.client.java.Tests.javaproj (hunk 2 above)
Comment (halter73) [reply src/SignalR/clients/java/signalr/test/signalr.client.java.Tests.javaproj]: The thing that makes Alpine very different from most other Linux distros is that it uses musl instead of glibc as its c standard library and they are not ABI compatible.

File: eng/targets/Helix.targets [hunk 3]
```diff
@@ -97,6 +97,7 @@
     <PropertyGroup>
       <BuildHelixPayload Condition="'@(_HelixApplicableTargetQueue->Count())' == '0'">false</BuildHelixPayload>
       <BuildHelixPayload Condition="'$(TargetArchitecture)' == 'arm64' AND '$(SkipHelixArm)' == 'true'">false</BuildHelixPayload>
+      <BuildHelixPayload Condition="$(HelixTargetQueue.StartsWith('(Alpine.')) AND '$(SkipHelixAlpine)' == 'true'">false</BuildHelixPayload>
```
Comment (halter73) [inline eng/targets/Helix.targets:100]: @BrennanConroy I'm continuing our conversation about the msbuild failure here.

The `<BuildHelixPayload Condition="$(SkipHelixQueues.Contains('$(HelixTargetQueue)'))">false</BuildHelixPayload>` a line below uses `'` in a condition. What are you recommending I change this line to?

Comment (BrennanConroy) [inline eng/targets/Helix.targets:100]: Might need '' around the condition
```suggestion
      <BuildHelixPayload Condition="'$(HelixTargetQueue.StartsWith('(Alpine.'))' AND '$(SkipHelixAlpine)' == 'true'">false</BuildHelixPayload>
```

Comment (halter73) [reply eng/targets/Helix.targets:100]: I was thinking that too, but then I saw the following which apparently doesn't need to be surrounded by quotes despite being very similar.

https://github.com/dotnet/aspnetcore/blob/219ecd6880120215137f4d6ef0dd2aa62c780af2/src/Components/WebAssembly/Sdk/src/targets/Microsoft.NET.Sdk.BlazorWebAssembly.Current.targets#L147

I wonder if we don't need to precent-encode the `(` though. Seems likely given the "Did you forget the closing parenthesis?" in the error.

```suggestion
      <BuildHelixPayload Condition="$(HelixTargetQueue.StartsWith('%28Alpine.')) AND '$(SkipHelixAlpine)' == 'true'">false</BuildHelixPayload>
```

**Comment by halter73** [comment]
I think this is good to go.

File: src/Testing/src/xunit/SkipOnAlpineAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+        public bool IsMet { get; } = !IsAlpine;
+
+        public string SkipReason => "Test cannot run on Alpine Linux.";
+
+        // This logic is borrowed from https://github.com/dotnet/runtime/blob/6a5a78bec9a6e14b4aa52cd5ac558f6cf5c6a211/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs
+        private static bool IsAlpine { get; } =
+            RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && File.Exists("/etc/os-release") &&
+            File.ReadAllLines("/etc/os-release").Any(line =>
+                line.StartsWith("ID=", StringComparison.Ordinal) && line.Substring(3).Trim('"', '\'') == "alpine");
```
Comment (dougbu) [inline src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Why can't this logic be included in `[SkipOnHelix]` to support `All.Alpine`❔ The inconsistency and the need for yet-another attribute are jarring.

**Review by BrennanConroy — APPROVED** [review]

**Review by jkotalik — APPROVED** [review]

**Review by dougbu — CHANGES_REQUESTED** [review]

File: src/Testing/src/xunit/SkipOnAlpineAttribute.cs (hunk 4 above)
Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Is it because its not specific to helix only?

Comment (halter73) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Unlike the base queue name, the name of the docker image doesn't show up in the "helix" environment variable `[SkipOnHelix]` uses so @MattGal suggested this. I prefer this anyway because it does the right thing even if someone is manually running tests on Alpine.

Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: We actually control setting the helix variable, but right in this case the queue name is a super unwieldy docker image :(  https://github.com/dotnet/aspnetcore/blob/master/eng/helix/content/RunTests/TestRunner.cs#L37

Comment (halter73) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Either way, I think this is better because it works even when not being run using a helix queue.

Comment (dougbu) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: If that's compelling (I'm not 💯% convinced but could be persuaded), we need to separate `All.OSX` from `[SkipOnHelix]` and put platform testing logic into a new `[SkipOnOSX]` or `[SkipOnMacOS]` attribute.

If we continue to create distinct test conventions, it'll keep getting harder to grasp how skip tests and avoid platform-specific aka queue-specific issues.

**Review by HaoK — APPROVED** [review]

File: src/Testing/src/xunit/SkipOnAlpineAttribute.cs (hunk 4 above)
Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Well if we are going that route, we probably should add an additional environment variable which is closer to a 'tag', so for OSX queues, it could say "helix;osx;linux" the alpine queue could be "helix;alpine;linux;docker" and then we could have something like `SkipOnTag(tag = "alpine")`, `SkipOnTag(tag = "osx")`

Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: We'd need to add/flow that metadata in the helix definitions too, I can do that in a separate PR I guess, I don't think @halter73 should need to do that as part of this PR.

Comment (dougbu) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: One way to describe my feeling here is holding out for a confirmed consistent direction on how we skip for specific platforms. Bringing everything, including the existing `All.OSX` and `[OSSkipCondition]` approaches together sounds like a great direction❕ Bonus points if you subsume `[MinimumOSVersion]` and `[MaximumOSVersion]` too 😺 

I'm :ok: with follow-up work in this direction as long as we agree and have some hope it'll be done soon. Any objection @halter73❔

Comment (dougbu) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Actually, why isn't this logic part of `[OSSkipCondition]`❔ Are we unable to extend the `OperatingSystems` to include narrower options❔

Comment (halter73) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: If this should be added to any existing attribute, it should be added to `[OSSkipCondition]`, but that only supports specifying "Linux", "MacOSX", "Windows". I don't know why `[SkipOnHelix]` has special `All.OSX` support given that `[OSSkipCondition]` exists, but that's not really relevant to this PR.

I didn't want to add full-blown distro support to `[OSSkipCondition]` since that could get pretty involved (see runtime's [PlatformDetection.Unix.cs](https://github.com/dotnet/runtime/blob/6a5a78bec9a6e14b4aa52cd5ac558f6cf5c6a211/src/libraries/Common/tests/TestUtilities/System/PlatformDetection.Unix.cs) for a taste), and I don't see us skipping any specific Linux distro other than Alpine. The only reason Alpine is special is because it uses musl instead of glibc like everey other distro. If it was easier to detect musl, I would do that.

@dougbu I need you to approve or remove your review before I can merge.

Comment (dougbu) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: > I need you to approve or remove your review before I can merge.

Yes, I know. Waiting to hear a confirmation @HaoK is willing to burn `All.OSX` with fire. If that happens and we can just use `[OSSkipCondition]` instead, I'll be happy because it keeps the concept count manageable.

We may still need to do some "stuff" for ARM / ARM64 skipping to make it consistent too but that's probably not going to fit into `[OSSkipCondition]` unless we add sub-categories like `Linux / Alpine` or `Windows / ARM64`.

Comment (halter73) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: This PR has nothing to do with OSX. I don't think it's worthwhile to combine `[SkipOnAlpine]` with any other attribute. There's only one existing test that needed this. I don't expect it to come up much.

Keeping the concept count manageable is a good goal, but this PR does not really increase the concept count in any appreciable way.

Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: So that complexity is there because not everything runs on helix, so OSSkip would disable it on non helix and helix, so skip on helix is basically an additional but orthogonal slice, since we had tests that only are skipped on helix + OSX for example, that were enabled on non helix OSX.  But now that almost all tests are on helix, we can probably just use OSSkip

Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: What I was proposing was closer to two concepts, we live OSSkip mostly alone, and modify SkipOnHelix to have traits so you can skip on things like osx/windows/arm.  There will be some overlap with OSSkip but the tags are specific to helix queues, which is something we need to be able to target (skipping on specific queues/types of queues)

Comment (dougbu) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: > There's only one existing test that needed this

Then why not just use this logic in the test itself❔

We may need to fix a general issue about skipping on specific fully-qualified queue names i.e. check against the un-split name. But, I'm less convinced now that we need a new attribute at all.

> What I was proposing was closer to two concepts

I'm not understanding how developers would choose between `[OSSkipCondition]` and `[SkipOnHelix(All.OSX)]` or your proposed `[SkipOnTag(tag: "osx")]`. Why do we need two ways to skip tests on macOS, Linux, or Windows❔

Put another way, why do we need to be able to skip these general categories of Helix queues❔

Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: OSSkipCondition is overly broad, generally tests seem to be failing on one queue at a time, like OSSkipLinux is probably not the right thing to do when its failing only on debian, or osx, or alpine for example

Comment (HaoK) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: Specifically OSSkip should be used if the test will never work on OSX, SkipOnHelix("osx") should be if it works on azdo runs but doesn't work on helix

Comment (dougbu) [reply src/Testing/src/xunit/SkipOnAlpineAttribute.cs:31]: > Specifically OSSkip should be used if the test will never work on OSX, SkipOnHelix("osx") should be if it works on azdo runs but doesn't work on helix

I get that but are any tests in the "build agent but not Helix agent" category these days❔

**Review by dougbu — APPROVED** [review]
Like the new approach @halter73. @HaoK let's discuss the Helix / OS concept count somewhere else&hellip;
